### PR TITLE
Vineyard data source and sink.

### LINKED ISSUE #1001: Add vineyard support
This issue is used to track the work about vineyard in the mars side. Currently the TODO list includes:

+ [x] use vineyard as data source and data sink (done by #1000)
+ [ ] implements a vineyard storage handler in mars
+ [ ] add a bypass to use vineyard to storage chunks and necessary meta

### PR DESCRIPTION
**Pull request by sighingnow**
## What do these changes do?

Mars use vineyard as datasource and data sink. The source code has already been cleaned up.

## Related issue number

Resolves #1001

### CODE REVIEW

**Comment by sighingnow** [comment]
This PR is ready for review. Details that the reviewer needs attention from reviewers are:

1. this PR is compatible with the plasma-based design, as you can see in CI, all tests works.
2. an option `vineyard.socket` is added
3. this PR do some hacks in operation schedulering and fusion, includes:
    + in the fusion part, operands that have an `expect_worker` attribute  won't be fused
    + in the assignment part, vineyard ops that has `expect_worker` attribute will be scheduled with respecting its `expect_worker`, even when the op has inputs.
4. the codacy complaints about the `pickle` related things but seems `noqa` doesn't work.

**Comment by sighingnow** [comment]
CI is green 😆

File: mars/tensor/datasource/from_vineyard.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        from ...session import Session
+
+        # ensure the input op has been evaluated
+        check_chunks_unknown_shape(op.inputs, TilesError)
+
+        if vineyard is None:
+            raise RuntimeError('vineyard is not available')
+        client = vineyard.connect(op.vineyard_socket)
+
+        ctx = get_context() or Session.default_or_local().context
```
Comment (qinxuye) [inline mars/tensor/datasource/from_vineyard.py:141]: Ditto.

File: mars/tensor/datastore/to_vineyard.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    def tile(cls, op):
+        from ...session import Session
+
+        check_chunks_unknown_shape(op.inputs, TilesError)
+
+        if vineyard is None:
+            raise RuntimeError('vineyard is not available')
+        client = vineyard.connect(op.vineyard_socket)
+
+        ctx = get_context() or Session.default_or_local().context
```
Comment (qinxuye) [inline mars/tensor/datastore/to_vineyard.py:128]: Ditto.

File: mars/dataframe/datasource/from_vineyard.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+    @classmethod
+    def tile(cls, op):
+        from ...session import Session
+
+        if vineyard is None:
+            raise RuntimeError('vineyard is not available')
+        client = vineyard.connect(op.vineyard_socket)
+
+        ctx = get_context() or Session.default_or_local().context
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply mars/dataframe/datasource/from_vineyard.py:82]: The `default_or_local().context` was learned from

https://github.com/mars-project/mars/blob/e04c6102c3fbece3828b416680478b62d70949b0/mars/learn/contrib/pytorch/dataset.py#L35

Should I remove that one in this PR as well?

[the remark this answers is not in the document]
Comment (qinxuye) [reply mars/dataframe/datasource/from_vineyard.py:82]: Oh, yes

File: mars/dataframe/datastore/to_vineyard.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    @classmethod
+    def execute(cls, ctx, op):
+        if vineyard is None:
+            raise RuntimeError('vineyard is not available')
+        client = vineyard.connect(op.vineyard_socket)
+        chunk_blob = vineyard.UInt64VectorBuilder(client, op.local_chunks).build(client)
+        ctx[op.outputs[0].key] = (client.instance_id, chunk_blob.id)
+
+
+class DataFrameToVineyard(DataFrameOperand, DataFrameOperandMixin):
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply mars/dataframe/datastore/to_vineyard.py:179]: Agree, will rename it.

**Comment by sighingnow** [comment]
CI is green again!

**Comment by sighingnow** [comment]
Thanks for merging!

**Pull request merged**